Add container scattering to Discus multiple scattering algorithm

DannyHindson opened the pull request:
**Description of work.**

Add support for scattering from a container to the Discus multiple scattering algorithm. This means that when generating a result for two scatters it will include extra situations in the simulation such as one scatter happening in the sample container and one in the sample.

The code changes involve converting various bits of processing to cope with a list of S(Q,w) instead of a single S(Q,w) for the sample. The algorithm class now has a new member called `m_SQWSs` that is a boost small_vector of S(Q,w) distributions.

**To test:**

Run the following script and check it works. I suggest you do it on a release build so it runs in a short time. It is a variation on one of the usage examples that is part of the documentation for the algorithm. The variation is that the input workspace has an Aluminium container set up as well as a sample. I've also changed the sample shape to be a cylinder instead of a sphere since it's easier to set up a container for a cylinder:
```
# import mantid algorithms, numpy and matplotlib
from mantid import mtd
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

# S(Q) consisting of single spike at q=1
X=[0.99,1.0,1.01]
Y=[0.,100,0.]
Sofq=CreateWorkspace(DataX=X,DataY=Y,UnitX="MomentumTransfer")

two_thetas=[]
for i in range(180):
    two_thetas.append(i)

# workspace with single bin centred at k=1 Angstrom-1
ws = CreateSampleWorkspace(WorkspaceType="Histogram",
                           XUnit="Momentum",
                           Xmin=0.5,
                           Xmax=1.5,
                           BinWidth=1.0,
                           NumBanks=len(two_thetas)//4,
                           BankPixelWidth=2,
                           InstrumentName="testinst")

ids = list(range(1,len(two_thetas)+1))
EditInstrumentGeometry(ws,
    PrimaryFlightPath=14.0,
    SpectrumIDs=ids,
    L2=[2.0] * len(two_thetas),
    Polar=two_thetas,
    Azimuthal=[90.0] * len(two_thetas),
    DetectorIDs=ids,
    InstrumentName="testinst")

SetSample(InputWorkspace=ws,
          Geometry={'Shape': 'Cylinder', 'Radius': 1.0, 'Height': 2.0, 'Centre': [0.,0.,0.], 'Axis': [1.0, 0.0, 0.0]},
          Material={'NumberDensity': 0.02, 'AttenuationXSection': 0.0,
                    'CoherentXSection': 0.0, 'IncoherentXSection': 0.0, 'ScatteringXSection': 80.0})

results_group = DiscusMultipleScatteringCorrection(InputWorkspace=ws, StructureFactorWorkspace=Sofq,
                                                   OutputWorkspace="MuscatResultsNoContainer", NeutronPathsSingle=1000,
                                                   NeutronPathsMultiple=10000, ImportanceSampling=True)

Scatter_1_DeltaFunctionNoContainer = CloneWorkspace('MuscatResultsNoContainer_Scatter_1')
Scatter_2_DeltaFunctionNoContainer = CloneWorkspace('MuscatResultsNoContainer_Scatter_2')
DeleteWorkspace('MuscatResultsNoContainer')

SetSample(InputWorkspace=ws,
          Geometry={'Shape': 'Cylinder', 'Radius': 1.0, 'Height': 2.0, 'Centre': [0.,0.,0.], 'Axis': [1.0, 0.0, 0.0]},
          Material={'NumberDensity': 0.02, 'AttenuationXSection': 0.0,
                    'CoherentXSection': 0.0, 'IncoherentXSection': 0.0, 'ScatteringXSection': 80.0},
          ContainerGeometry={'Shape': 'HollowCylinder', 'Height': 2.0,
                  'InnerRadius': 1.0, 'OuterRadius': 1.2, 'Center': [0.,0.,0.], 'Axis': [1.0, 0.0, 0.0]},
          ContainerMaterial={'ChemicalFormula': 'Al',
                  'NumberDensity': 0.1})

results_group = DiscusMultipleScatteringCorrection(InputWorkspace=ws, StructureFactorWorkspace=Sofq,
                                                   OutputWorkspace="MuscatResultsWithContainer", NeutronPathsSingle=1000,
                                                   NeutronPathsMultiple=10000, ImportanceSampling=True)
# Can't index into workspace group by name (yet) so just get the members from the ADS instead
Scatter_1_DeltaFunctionWithContainer = CloneWorkspace('MuscatResultsWithContainer_Scatter_1')
Scatter_2_DeltaFunctionWithContainer = CloneWorkspace('MuscatResultsWithContainer_Scatter_2')
DeleteWorkspace('MuscatResultsWithContainer')

msplot = plotBin('Scatter_2_DeltaFunctionNoContainer',0)
msplot = plotBin('Scatter_2_DeltaFunctionWithContainer',0, window=msplot)
```
The plot should show two simulated intensities that are broadly similar. The intensity with a container is slightly smaller because there's more material in the beam.

Fixes #33694.

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson committed:
Initial work on supporting container scattering

Modify the calculation of the path length in updateWeightAndPosition
Rework the QSS normalisation to be done per S(Q) rather than using a
global QSS factor at the end of simulatePaths
Also reduce number of sqrt calls from 2 to 1 in interpolateSquareRoot

DannyHindson committed:
Widen tolerance for unit test

The changes in an earlier commit to switch the elastic input units
from wavelength to momentum had shifted the actual value so that it
was very close to the 1e-05 tolerance. This was due to the bin order
being switched and the test bin became [2] instead of [0]. This in
turn meant the random numbers used for bin [2] were different.

The last commit on this change to support containers moved it by a tiny
amount and it has dropped over the 1e-05 threshold. Note - the reference
value is still correctly centred it's just the spread is bigger than
originally thought

DannyHindson committed:
Improve normalisation scheme ready for >1 S(Q,w)

Change q_dir so the behaviour without importance sampling is better
Integration is now over a rectangular area without retries if the
sampled Q,w is outside the kinematic range. In this case the weight
is zero rather than a retry to get a valid Q,w

DannyHindson committed:
Change normalisation scheme

Don't scale the S(Q) any more
Just normalise the integral over QS(Q) in the multiple scatters so
that the integral is a fraction of 2k^2. S(Q) used in final leg is
the supplied\unscaled S(Q)

DannyHindson committed:
Update test reference values for modified normalisation scheme

DannyHindson committed:
Couple of changes

Remove returnCumulative switch. Doesn't seem necessary having a separate code path to test\support
Remove validation that S(Q,w) has some values for negative w because q_dir now supports tracks that run out of energy

DannyHindson committed:
Rework the normalisation of the Q.S(Q) integral (again!)

The integrals over the kinematic range of a S(Q,w) aren't v
accurate due to the use of a regular set of bins for all Q
and possibility of truncation at large w
eg for a QENS Lorentzian this means the binning is too wide at low and
the tails are chopped because w doesn't extend to infinity
So don't try and calculate the effective cross section by integrating
Q.S(Q,w) over all Q,W. Instead get the user to supply a cross section
as a function of k and normalise the integral at a particular k
by dividing by the actual integral over the kinematic range - both the
numerator and denominator will be similarly afflicted by any problems
in S(Q,w) from binning\truncation
Note - getting user to supply cross section as function of k was an
original Discus feature. I'd not understood the value of this until now

DannyHindson committed:
Further work on supporting containers

Simplify code in UpdateWeightAndPosition. Total weight doesn't depend
on where the scatter happens so can do this in the first loop
Also remove some redundant calls to new_vector

DannyHindson committed:
Further tidy up

Return scattering cross section from UpdateWeightAndPosition instead
of component id.
Start changes to support multiple S(Q) distributions

DannyHindson committed:
More work on handling multiple S(Q,w) inputs

DannyHindson committed:
Some bug fixes from testing on new container enabled code

DannyHindson committed:
More bug fixes\improvements to container version

DannyHindson committed:
More work on supporting >1 S(Q,w)

Start setting up vector of S(Q,w) and its derived versions with one
per material. Perhaps easier than one per component because the
components aren't often named whereas the materials are

DannyHindson committed:
Finish off changes to have a per material list of S(Q,w)

DannyHindson committed:
Sample w based on w values rt index in case bins aren't same width

DannyHindson committed:
Fix couple of glitches

validateInputs crashed if InputWorkspace was a group workspace
If QSQIntegral was zero the q_dir logic gave NaN values

DannyHindson committed:
Incorporate performance improvements from profiling

DannyHindson committed:
Performance enhancement

Reduce number of points in profiles that store QSQ integrals
Extra points are added to QSQ to make integration of QSQ more accurate
but where we're storing a cumumlative integral we don't need the integral
value itself at all these extra Q's - perhaps slows down lookups using
std::upper_bound later on
Also adjust code using epsilon to scale it properly and use std::next_after
where possible

DannyHindson committed:
More performance improvements

Use boost small_vector for vector of SQ-related workspaces
Use string_view instead of string
Also revert temporary change to use interpolateFlat instead of
interpolateGaussian (temp change was in place to make it easier
to compare results with and without IS)

DannyHindson committed:
Stop dividing by QSQIntegral of zero

DannyHindson committed:
Add info message when defaulting S(Q,w) to isotropic

DannyHindson committed:
Don't add duplicate materials to m_SQWSs

DannyHindson committed:
Move to workspace list per component rather than per material

To assist in logging the scatter point counts per component
Currently looking up component from the track intersects by
matching the IObject pointer
Maybe move to hash of this in future? Or perhaps set the id
on the IObjects and use that (eg "Sample", "Env1", "Env2" etc)

DannyHindson committed:
Fix problem in maths in weight calculation relating to t1

The weight calculation associated with the t1 variable was wrong in
case where there's a container. The wrong calculation reduced to the
correct weight in the case of no container so wasn't v obvious
Spotted this because results for v thin container weren't approximately
the same as sample only

DannyHindson committed:
Fix problem with Interpolate2D

...when running inelastic calculation on a S(Q,w) that is elastic
ie zero apart from a w=0. The hidden width of +/0.5 was meaning
S(Q,w) was giving non-zero values for w!=0
Also modify prepareQSQ to not add extra points if maxQsupplied=0.

DannyHindson committed:
Remove some nastiness from QSQ integration

Calculate QSQ at the supplied Q points rather than basing
this off the enhanced QSQ profile that includes a qmax (~DBL_MAX)
and some extra points (that needed to be stripped out)
Nature of flat interpolation means the integral at k=qmax/2
will never be used

DannyHindson committed:
Add parameter to allow S(Q,w) to be normalized

For use in calculations where more than one material involved
and S(Q,w) may not be normalized in all cases
Also fix problem where algorithm run from C++ test multiple times
and m_SQWSs wasn't being cleared between runs

DannyHindson committed:
Recentre some test ref values following rebase onto main

Fix to flip sign of flat plate seems to have collided with
change to tighten tolerance on the first direct test on
realistic structure factor
Keep the ref values for the first three tests the same because
they're essentially the same calculation done in 3 diff ways

DannyHindson committed:
Add some missing omega terms into the documentation

...ahead of reworking to include the description of the
container version of the algorithm

DannyHindson committed:
Various small code changes and started docs update

Docs update to include new S(Q,w) normalisation approach required for
sample + container calculation
Some doc updates to clarify what's currently in main also - perhaps remove
some of this content before finalising the sample+container version but
leave in for now so have correct documentation for the main logic also
Adjust the "NormalizeStructureFactors" logic to swap in the k dependent
cross section for the final leg

DannyHindson committed:
Add documentation on container enabled calculation

DannyHindson committed:
Replace mu unit with pt

mu not recognized in qtqhelp docs for some reason

DannyHindson committed:
Couple more Discus related doc updates

DannyHindson committed:
Fix problem with shadow definition of i

DannyHindson committed:
Add test on structure factor list supplied

While doing this noticed that a group in the StructureFactorWorkspace
parameter was triggering the process groups behaviour in algorithm
So disable this
Also fixed bug in validateInputs where component id rather than material
was being input into a list

DannyHindson committed:
Add validation on whether material has been set

On sample and any environment components
While doing this make sure it can run on a container only set up
Fix some more missing initializer warnings

DannyHindson committed:
Fix compiler warning about int types

DannyHindson committed:
Fix gcc compiler warning

DannyHindson committed:
Speed up integrateCumulative

Do this to stop SimpleShapeDiscusInelastic unit test timing out
Perhaps needs further work - perhaps calculating the QSQ integrals
on demand should be better given that I'm currently redoing
the integral per bin. The number of k's that the integral is
required for per bin won't be that many

DannyHindson committed:
Rework the I(k) calculation to improve performance

Calculate the I(k) on demand and cache the result rather than
pre-calculating on arbitrary set of k values and interpolating.
This will speed up the performance for nscatters=2 or 3 where the
number of k values where I(k) is required for a specific bin is small
eg if running with NormalizeStructureFactors=False and nscatters=2
then I(k) is only needed for a single k value
The situation where the largest number of I(k) calculations will be
required will probably be inelastic where k changes after each scatter
but the current approach of w being one of a discrete set of values
means the k values where I(k) is required will also be a discrete set
of finite\modest size
Separate cache per spectrum so that can reuse results across bins will
benefit direct calculations where initial k is same for all bins

DannyHindson committed:
Pull out the component lookup into a function to remove duplication

DannyHindson committed:
Fix cppcheck warnings and reinstate concave comment

DannyHindson committed:
Fix compilation problem on Ubuntu non-conda

DannyHindson committed:
Add release note

DannyHindson committed:
Remove spurious change to ISISIndirectEnergyTransfer

DannyHindson committed:
Two minor changes

Reinstate virtual destructor. Must have got dropped during rebase
Change the equal bins validation to be a common bins validation in case
where inelastic input workspace has deltaE along x axis

DannyHindson committed:
Pull some of the new code out of exec and tidy up

sf1919 commented:
Looking at the code I note that ```DiscusMultipleScatteringCorrection``` has a new property - ```NormalizeStructureFactors```. Is there a way to test this as I don't think the test above covers this?

DannyHindson commented:
> Looking at the code I note that `DiscusMultipleScatteringCorrection` has a new property - `NormalizeStructureFactors`. Is there a way to test this as I don't think the test above covers this?

Yes good point. You can just add `, NormalizeStructureFactors=True` to the two calls to the `DiscusMultipleScatteringCorrection`  algorithm in the test script. The default value of this new parameter is False so this will trigger the new functionality. I think the two spectra that get plotte (with and without a can) should still be similar. I think the y values of the two spectra will probably be slightly different to when you run the script as is which effectively has `NormalizeStructureFactors=False`

sf1919 commented on docs/source/algorithms/DiscusMultipleScatteringCorrection-v1.rst:
```diff
@@ -115,6 +124,40 @@ The importance sampling has also been implemented for inelastic instruments by f
+Support for sample environment
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The calculation can include scattering from the sample environment (eg can) in the Monte Carlo simulation. The term "segment" has previously been used to refer to a straight neutron path between two scattering events. For the purpose of this description the term "link" will be used to refer to a subsection of a segment that lies within a single material.
```
```suggestion
The calculation can include scattering from the sample environment (e.g. can) in the Monte Carlo simulation. The term "segment" has previously been used to refer to a straight neutron path between two scattering events. For the purpose of this description the term "link" will be used to refer to a subsection of a segment that lies within a single material.
```

sf1919 commented on docs/source/algorithms/DiscusMultipleScatteringCorrection-v1.rst:
```diff
@@ -115,6 +124,40 @@ The importance sampling has also been implemented for inelastic instruments by f
+
+The calculation can include scattering from the sample environment (eg can) in the Monte Carlo simulation. The term "segment" has previously been used to refer to a straight neutron path between two scattering events. For the purpose of this description the term "link" will be used to refer to a subsection of a segment that lies within a single material.
+
+The modified calculation is illustrated here with an example of a sample contained in a can where a track may contain three different links (can then sample then can). If the selected scatter point occurs somewhere in the third link, the quantity :math:`t_i` is redefined as:
```
```suggestion
The modified calculation is illustrated here with an example of a sample contained in a can where a track may contain three different links (can, then sample, then can). If the selected scatter point occurs somewhere in the third link, the quantity :math:`t_i` is redefined as:
```

sf1919 commented on docs/source/release/v6.5.0/Framework/Algorithms/New_features/34203_AddContainerScatteringToDiscus.rst:
```diff
@@ -0,0 +1 @@
+- The algorithm :ref:`DiscusMultipleScatteringCorrection <algm-DiscusMultipleScatteringCorrection>` now simulates scattering in the a container (and more generally in any sample environment components) if a user sets up the shapes\materials in the input workspace prior to running the algorithm
```
```suggestion
- The algorithm :ref:`DiscusMultipleScatteringCorrection <algm-DiscusMultipleScatteringCorrection>` now simulates scattering in the a container (and more generally in any sample environment components) if a user sets up the shapes\materials in the input workspace prior to running the algorithm.
```

SpencerHowells replied:
Remove 'the' or  'a' in 'in the a container'?

sf1919 reviewed (changes requested):
One or two very minor changes and some queries to better understand the code.

DannyHindson committed:
Performance improvement in integrateCumulative

Use getSpectrum(i) instead of histogram(i) so that reference to histogram
returned instead of copy
Add option for integrateCumulative to only return total integral
Spotted slowness when running calculations on indirect geometry
instrument (OSIRIS)

DannyHindson committed:
Revert I(k) calculation approach

In order to fully resolve the performance problem observed running calcs
on OSIRIS data with normalizestructurefactors=true
Revert to approach of pre-calculating I(k) on a few important values
and then interpolating as required rather than calculating on demand
Some of the code previously removed in commit 440e9369c6bd95d0fe107ced192c9dd216aad7b1
has been reinstated

DannyHindson commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -172,65 +178,97 @@ std::map<std::string, std::string> DiscusMultipleScatteringCorrection::validateI
+  size_t nEnvComponents = 0;
+  if (inputWS->sample().hasEnvironment())
+    nEnvComponents = inputWS->sample().getEnvironment().nelements();
+  if (inputWS->sample().getShape().hasValidShape())
```
I can't reuse atLeastOneValidShape here because I need to know the exact number of valid shapes. If there are 3 valid shapes for example I need to step through the 3 shapes and check each one has a valid material set up.
Having said that I will make a small change to tidy the code up here. I should be calling hasValidShape on all the possible shapes that are present whereas I only currently do it on the sample shape. I need to additionally call hasValidShape on the sample environment shapes which are checked in the loop starting on L206

DannyHindson commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -172,65 +178,97 @@ std::map<std::string, std::string> DiscusMultipleScatteringCorrection::validateI
+      if (SQWS->getAxis(1)->isSpectra())
+        issues["StructureFactorWorkspace"] += "S(Q, w) must have a numeric spectrum axis\n";
+      std::vector<double> wValues;
+      if (SQWS->getAxis(0)->unit()->unitID() == "DeltaE") {
+        if (!SQWS->isCommonBins())
+          issues["StructureFactorWorkspace"] += "S(Q,w) must have common w values at all Q";
```
So the original limitation on needing some negative values wasn't a user requirement. It was more a convenience thing because it made the code simpler to require this (so perhaps laziness on my part). 
A couple of things have changed since I put this limitation in place:

1) The work in this PR meant that the simplifications I was trying to hold on to couldn't be carried forward
2) Spencer has some S(q,w) datasets that don't have any values for w<0 so the algorithm wasn't generating a result for those even though a meaningful result could have been created

So I've removed the validation and it all seems to work fine. I'll mention this in the release note

DannyHindson commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -686,102 +768,151 @@ DiscusMultipleScatteringCorrection::generateInputKOutputWList(const double efixe
+ * variable, the w values correspodning to each value of the pseudo variable
  */
-void DiscusMultipleScatteringCorrection::prepareCumulativeProbForQ(double kinc, const MatrixWorkspace_sptr &PInvOfQ) {
+std::tuple<std::vector<double>, std::vector<double>, std::vector<double>>
```
Yes it's the return type of the function. Because it's quite long the clang formatting has put it on its own line.

DannyHindson commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -1212,76 +1386,100 @@ std::tuple<double, double> DiscusMultipleScatteringCorrection::getKinematicRange
+    // only calculated for double scattering and easier to normalise in that case
+    double wRange;
+    /*// the energy transfer must always be less than the positive value corresponding to energy going from ki to 0
+    // Note - this is still the case for indirect because on a multiple scatter the kf isn't kfixed
```
`fkixed` is actually a quantity that's used by the algorithm so this is OK rather than a typo

DannyHindson commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -1212,76 +1386,100 @@ std::tuple<double, double> DiscusMultipleScatteringCorrection::getKinematicRange
+    double wRange;
+    /*// the energy transfer must always be less than the positive value corresponding to energy going from ki to 0
+    // Note - this is still the case for indirect because on a multiple scatter the kf isn't kfixed
+    double wMax = fromWaveVector(kinc);
+    // find largest w bin centre that is < wmax and then sample w up to the next bin edge
+    auto it = std::lower_bound(wValues.begin(), wValues.end(), wMax);
+    int iWMax = static_cast<int>(std::distance(wValues.begin(), it) - 1);*/
```
Yes they are but the code would benefit from a further comment here about why. I'll add one in. The code that's commented out here and also on L1452 is an optimisation that should mean you can get more accurate results but I didn't want to introduce it at this stage because I've spent a long time on validating the results without this. So in the spirit of taking things one step at a time I was planning on raising a separate PR for this

DannyHindson commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -1212,76 +1386,100 @@ std::tuple<double, double> DiscusMultipleScatteringCorrection::getKinematicRange
+      wBinEdges.reserve(wValues.size() + 1);
+      VectorHelper::convertToBinBoundary(wValues, wBinEdges);
+      // w bins not necessarily equal so don't just sample w index
+      wRange = /*std::min(wMax, wBinEdges[iWMax + 1])*/ wBinEdges.back() - wBinEdges.front();
```
See previous comment

DannyHindson committed:
Incorporate review comments

Tidy up the shape validation in the validateInputs function
Pull out the sampleQW logic for the code branch that doesn't use
importance sampling into its own function (was like this in the
code without container support)
Few misc documentation changes
Add doxygen documentation to q_dir and update doxygen comments for
integration functions

sf1919 reviewed (approved):
The code works well and I'm satisfied the changes I requested have been made where possible. Where not possible this has been explained fully.

There is mention of further work - I don't think an issue has been raised for this? If not I think it would be good to capture this via a new issue.

Happy to approve.
